### DISCUSSION TITLE
(newbie) Clarifying some things

### DISCUSSION THREAD

**Discussion by yatesco** [schell/mogwai#95] [Q&A] [answered]
Hi there - I hope you don't mind a bunch of questions, and it's entirely possible I've misunderstood something  ;-)

### Controlled inputs
In react inputs (e.g. text field) are either controlled or uncontrolled. Are inputs here always controlled, so each keypress transmits and the actual rendered input is just a fold over the received transmissions? Or are they uncontrolled? How is validation handled in each scenario?

### Json from server
When new state comes from the server, I assume you update the stateful struct of the view, but how do you signal a redraw of the view? Would you have a synthetic "Redraw" that you would transmit? In the Clicks example, we would update the clicks to the new value, but the actual View is a fold over the individual transmissions. I don't get how to say "stop folding over the incremental txs - here is a whole new complete state".

### Accessing the HTMLElement itself
If I wanted to use Highcharts, for example, I would need access to the underlying div. Is this possible? 

I think that's it for now. `mogwai` looks ace, and I'm very excited about diving in, I just wanted to sanity check my assumptions first :-)

**Comment by schell** [comment, marked as answer]
First off, thanks for you interest! I'll answer your questions as best I can. 

> Are inputs here always controlled ... or are they uncontrolled? 

I think whether they are controlled or uncontrolled (in React terms) depends on how the view is set up. 
`mogwai`'s views are quite a bit different since you can fold and scan (in FP terms) at the view definition site (inside the `ViewBuilder`). But the main logic loop is also a fold, however implicitly so. 

It sounds like you're asking a more general question about a view's internal state. My advice is to keep views just dumb pipes that receive messages and display them with as little interpolation and state as possible. The One True State™ should be kept in your logic loop. Whenever that state changes you explicitly send a message out to the view to update. Your logic loop can receive messages from outside that update the One True State™ from some upstream component and then subsequently send that change as a message out to its own view as previously stated. So I think in this way you could have uncontrolled or controlled values.  

The `0.6.0` branch is going to help with this management immensely using `Relay`, `Input`, `InputFan`, `Output` and `Captured`. I'd suggest tracking that branch since it has breaking changes and solves lots of problems with the `0.5` release. 

Here's a quick example (thanks to my collegue @rmfisher). Given this react blurb that shows a controlled input:
```js
[inputValue, setInputValue] = useState('')

return <input value={inputValue} onChange={e => setInputValue(e.target.value)} />
```

We could do the equivalent in `mogwai` (with a lot of extra typechecking):
```rust 
/// Here's a struct for our view. Technically we call this a "relay" because it's a struct that "relays" messages
/// to and from a view.
#[derive(Clone)]
struct MyInput {
    dom: Captured<Dom>,
    state: Arc<mogwai::core::futures::lock::Mutex<String>>,
    value: Input<String>,
    on_change: Output<String>,
}

impl MyInput {
    pub fn new(init: impl Into<String>) -> Self {
        Self {
            state: Arc::new(Mutex::new(init.into())),
            value: Input::default(),
            on_change: Output::default(),
        }
    }
    pub async fn update_the_one_true_state(&self, new_value) {
        let mut state = self.state.lock().await;
        *state = new_value;
        self.value.set(&state).await.unwrap();
    }
}

/// This trait implementation allows us to define the view and its logic in one place.
impl Relay<Dom> for MyInput {
    type Error = anyhow::Error; 
    
    fn view(&mut self) -> ViewBuilder<Dom> {
        builder! { 
            <input value=("", self.input.stream().unwrap()) on:change=self.on_change.contra_filter_map(|ev:DomEvent| {
                let ev:web_sys::Event = ev.browser_event()?;
                let target = ev.target()?;
                let input = dyn_ref::<web_sys::HtmlInputElement>()?;
                Some(input.value())
            }
                capture:view=self.dom.sink()>
            </input>
        }
    }
    
    fn logic(self) -> Pin<Box<dyn Spawnable<Result<(), Self::Error>>>> {
        Box::pin(async move {
            while let Some(new_value) = self.on_change.get().await {
                self.update_the_one_true_state(new_value).await;
            }
        })
    }
}
```
Given that definition you could use it this way:
```rust
let input = MyInput::new("hello there!");

// it is Clone + Send + Sync so we can send off a clone of it to be built, which will use the Relay impl to construct actual browser DOM. 
let View{ inner: dom } = input.clone().into_view().unwrap();
dom.run().unwrap();

// We can still interact with the built view (which has been handed off to the browser).
input.set_the_one_true_state("i see what you did there").await;

// We can even get a clone of the DOM node back:
let dom = input.dom.get().await;

// When we're finally done and want it all to go away we can detach the dom by hand.
input.set_the_one_true_state("goodbye, cruel world!").await;
dom.detach();
```

> If I wanted to use Highcharts, for example, I would need access to the underlying div. Is this possible?

Yes! In the example above you can see the `DomEvent` type getting broken down into its underlying `web-sys` type. The same thing can be done if you have a `Dom`. See [`Dom::clone_as`](https://docs.rs/mogwai/0.5.3/mogwai/view/struct.Dom.html#method.clone_as) and [`Dom::visit_as`](https://docs.rs/mogwai/0.5.3/mogwai/view/struct.Dom.html#method.visit_as) for more info. 

> mogwai looks ace, and I'm very excited about diving in...

Thank you so much! I highly suggest starting off on the `0.6.0` branch if you can manage it. 

I think this shows that `mogwai` is very flexible. You have a lot of options for how you organize your UI graph. `Relay`s are what help me the most. They are a bit boilerplate-y but they offer the best semantics. Whatever you choose - the goal of `mogwai` is to be a layer in your application stack - not the overseer or all-encompassing framework. 

Let me know you thoughts and happy hacking! ☕

**Reply by schell** [reply]
Also, instead of the `Arc<Mutex<String>>` in the above example, you could simplify things quite a bit using (in `0.6`) `mogwai::core::model::Model`. `Model` is essentially an `Arc<Mutex<T>>` that sends updates downstream automatically. Here's the same thing above rewritten with a model:

```rust
/// Here's a struct for our view. Technically we call this a "relay" because it's a struct that "relays" messages
/// to and from a view.
#[derive(Clone)]
struct MyInput {
    dom: Captured<Dom>,
    state: Model<String>,
    value: Input<String>,
    on_change: Output<String>,
}

impl MyInput {
    pub fn new(init: impl Into<String>) -> Self {
        Self {
            state: Model::new(init.into()),
            value: Input::default(),
            on_change: Output::default(),
        }
    }
    pub async fn update_the_one_true_state(&self, new_value) {
        self.state.visit_mut(|s| *s = new_value).await; // no need to send updates by hand, the latest value will be sent automatically
    }
}

/// This trait implementation allows us to define the view and its logic in one place.
impl Relay<Dom> for MyInput {
    type Error = anyhow::Error; 
    
    fn view(&mut self) -> ViewBuilder<Dom> {
        builder! { 
            <input value=(self.state.current().unwrap(), self.state.stream()) on:change=self.on_change.contra_filter_map(|ev:DomEvent| {
                let ev:web_sys::Event = ev.browser_event()?;
                let target = ev.target()?;
                let input = dyn_ref::<web_sys::HtmlInputElement>()?;
                Some(input.value())
            }
                capture:view=self.dom.sink()>
            </input>
        }
    }
    
    fn logic(self) -> Pin<Box<dyn Spawnable<Result<(), Self::Error>>>> {
        Box::pin(async move {
            while let Some(new_value) = self.on_change.get().await {
                self.update_the_one_true_state(new_value).await;
            }
        })
    }
}
```

**Reply by yatesco** [reply]
thanks!

**Comment by schell** [comment]
Take note that most of this info is now out of date as of the 0.7 release.
